roscala/src/main/scala/coop/rchain/rosette/Ctxt.scala

```diff
@@ -1,42 +1,102 @@
 import scala.collection.mutable
 
-case class Ctxt(argvec: Tuple,
-                code: Code,
-                ctxt: Ctxt,
-                env: Ob,
-                monitor: Monitor,
+case class Ctxt(tag: Location,
                 nargs: Int,
                 outstanding: Int,
                 pc: PC,
-                reg: Seq[Ob],
-                rslt: Ob,
-                trgt: Ob,
-                selfEnv: Ob,
-                override val _slot: mutable.Seq[Ob],
-                tag: Location)
+                rslt: Ob, // reg[0]
+                trgt: Ob, // reg[1]
+                argvec: Tuple, // reg[2]
+                env: Ob, // reg[3]
+                code: Code, // reg[4]
+                ctxt: Ctxt, // reg[5]
+                self2: Ob, // reg[6]
+                selfEnv: Ob, // reg[7]
+                rcvr: Ob, // reg[8]
+                monitor: Monitor, // reg[9]
+                override val _slot: mutable.Seq[Ob])
     extends Ob {
+  def applyK(result: Ob, tag: Location)(state: VMState): (Boolean, VMState) =
+    ctxt.rcv(result, tag)(state)
+
   def arg(n: Int): Option[Ob] = argvec.elem.lift(n)
 
+  def getReg(r: Int): Option[Ob] =
+    r match {
+      case 0 => Some(rslt)
+      case 1 => Some(trgt)
+      case 2 => Some(argvec)
+      case 3 => Some(env)
+      case 4 => Some(code)
+      case 5 => Some(ctxt)
+      case 6 => Some(self2)
+      case 7 => Some(selfEnv)
+      case 8 => Some(rcvr)
+      case 9 => Some(monitor)
+      case _ => None
+    }
+
   /** This is necessary because the compiler sometimes arranges to
     *  provide an argvec that is acually longer than nargs indicates. If
     *  we are about to expose the context to the outside world, we need
     *  to clean it up so that it appears consistent.
     */
   def prepare(): Ctxt = this.copy(argvec = argvec.makeSlice(0, nargs))
 
-  def ret(rslt: Ob): Boolean = true
+  def rcv(result: Ob, loc: Location)(state: VMState): (Boolean, VMState) = {
+    val storeResult = Location.store(loc, this, state.globalEnv, result)
+
+    storeResult match {
+      case StoreCtxt(newCtxt) =>
+        val newState = state
+          .set(_ >> 'ctxt >> 'ctxt)(newCtxt)
+          .update(_ >> 'ctxt >> 'ctxt >> 'outstanding)(_ - 1)
+          .updateSelf(st =>
+            if (st.ctxt.ctxt.outstanding == 0) scheduleStrand(st) else st)
+
+        (false, newState)
+
+      // TODO: Add other store cases
+      case _ =>
+        (true, state)
+    }
+  }
+
+  def ret(result: Ob)(state: VMState): (Boolean, VMState) =
+    if (tag != LocLimbo) {
+      applyK(result, this.tag)(state)
+    } else {
+      (false, state)
+    }
 
   def scheduleStrand(state: VMState): VMState =
     state.update(_ >> 'strandPool)(_ :+ this)
 
-  def vmError(state: VMState): Ob = {
+  def setReg(r: Int, ob: Ob): Option[Ctxt] =
+    r match {
+      case 0 => Some(copy(rslt = ob))
+      case 1 => Some(copy(trgt = ob))
+      case 2 => Some(copy(argvec = ob.asInstanceOf[Tuple]))
```

Comment:
Since almost everything in Rosette takes and returns `Ob`, I think there is not really a way to avoid it. 
However I think that there are a few places where we will figure out that we can make things more precise.